test_util/cluster.py

```diff
@@ -0,0 +1,292 @@
+import itertools
+import logging
+import os
+import stat
+from contextlib import contextmanager
+
+import passlib.hash
+import pkg_resources
+from retrying import retry
+
+import gen.calc
+import test_util.installer_api_test
+import test_util.test_runner
+from ssh.ssh_tunnel import SSHTunnel, TunnelCollection
+
+
+class Ssher:
+
+    def __init__(self, user, home_dir, key_path):
+        self.user = user
+        self.home_dir = home_dir
+        self.key_path = key_path
+
+        try:
+            os.chmod(self.key_path, stat.S_IREAD | stat.S_IWRITE)
+        except FileNotFoundError as exc:
+            raise Exception('SSH key not found at path {}'.format(self.key_path)) from exc
+        except OSError as exc:
+            raise Exception('Unable to set permissions on SSH key at {}: {}'.format(
+                self.key_path,
+                exc,
+            )) from exc
+
+    def tunnel(self, address):
+        return SSHTunnel(self.user, self.key_path, address)
+
+    @contextmanager
+    def tunnels(self, addresses):
+        hostports = [a + ':22' for a in addresses]
+        with TunnelCollection(self.user, self.key_path, hostports) as tunnel_collection:
+            yield tunnel_collection.tunnels
+
+    def remote_cmd(self, addresses, cmd):
+        with self.tunnels(addresses) as tunnels:
+            for tunnel in tunnels:
+                tunnel.remote_cmd(cmd)
+
+
+class Cluster:
+
+    def __init__(
+        self,
+        ssh_info,
+        ssh_key_path,
+        provider,
+        masters,
+        agents,
+        public_agents,
+        bootstrap_host=None,
+        cleanup_callback=None,
+    ):
+        # Assert that this is a valid cluster.
+        gen.calc.validate_num_masters(str(len(masters)))
+        assert all(len(x) >= 0 for x in [agents, public_agents])
```

Comment:
Actually, since the integration tests are a separate thing, this shouldn' actually check either agents or public_aggents have anything. Having 0 public agents is fine.